File: DSA/Java/src/dsa/lib/algo/sort/SelectionSort.java
package dsa.lib.algo.sort;

import dsa.lib.utils.Range;

public class SelectionSort {
    public static void sort(int[] ary) {
        int n = ary.length;
        for (int i = 0; i < n - 1; i++) {
            int minPos = arrayMinPos(ary, i, n);
            Range.swap(ary, i, minPos);
        }
    }

    /**
     * Find the index of the minimum element from the range of array
     *
     * @param ary The whole array
     * @param st  The start position of the range, inclusive.
     * @param end The end position of the range, exclusive.
     * @return The index of the minimum element
     */
    public static int arrayMinPos(int[] ary, int st, int end) {
        int m = st;
        for (int i = st + 1; i < end; ++i) {
            if (ary[i] < ary[m]) m = i;
        }
        return m;
    }
}

File: DSA/Python/src/dsa/lib/algo/sort/selection_sort.py
from typing import List

from dsa.lib.utils import arrays


class SelectionSort:
    def sort(self, ary: List[int]) -> None:
        n = len(ary)
        for i in range(n - 1):
            min_idx = self.array_min_pos(ary, i, n)
            arrays.swap(ary, i, min_idx)

    @staticmethod
    def array_min_pos(ary: List[int], st: int, end: int) -> int:
        """
        Find the position of minimum element in the range of array

        :param ary: The whole array.
        :param st: The start position of the range array, inclusive.
        :param end: The end position of the range array, exclusive.
        :return: The position of the minimum element.
        """
        m: int = st
        for i in range(st + 1, end):
            if ary[i] < ary[m]: m = i
        return m
